### PR TITLE
[MRG+1] Fixed bug FormRequest.from_response() clickdata ignores input[type=image]

### PR DESCRIPTION
**Pull request by virmht**
BugFix for issue #3139

### CODE REVIEW

**Comment by virmht** [comment]
@cathalgarvey , please review the PR. Thanks :)

**Comment by cathalgarvey** [comment]
Thanks @virmht I'll give it a look :)

**Comment by cathalgarvey** [comment]
Looking at this, I think it's correct. And it includes a test-case.

I'm not sure though whether the XPath should be reformatted to use a single regex pattern, matching `^(submit|image)$` for example, or whether we should accept that the list of OR predicates may grow, and reformat the Xpath expression to keep one OR option per line?

E.g., the difference between a DRYer single-pattern match for images and submits:

```python
form.xpath(
    'descendant::*[(self::input or self::button)'
    ' and re:test(@type, "^(submit|image)$", "i")]'
   '|descendant::button[not(@type)]',
    namespaces={"re": "http://exslt.org/regular-expressions"})
```
..and with one-option-per-line, keeping options clearly separate to assist debug..
```python
form.xpath(
    'descendant::*[(self::input or self::button) and re:test(@type, "^submit$", "i")]'
    '|descendant::*[(self::input or self::button) and re:test(@type, "^image$", "i")]'
    '|descendant::button[not(@type)]',
    namespaces={"re": "http://exslt.org/regular-expressions"})
```

@kmike - I know you care about DRY, but Xpath expressions will be missed by linters and other automated tools, so maybe clarity in each case is a useful way to keep it clean? What do you think?

@virmht - What's your opinion? :)

**Comment by cathalgarvey** [comment]
(With or without changes, though, consider this MRG+1 from me - I'll change the title to reflect)

**Comment by virmht** [comment]
@cathalgarvey, I get your point. Which pattern to follow depends on the expected length of the list of OR predicates. If the list size doesn't increase by more than 3-4 then, maybe, we will be just fine with using a single regex pattern, otherwise, using the one-predicate-per-line pattern will be a better choice. So, how long, do you think, can the list get ?

File: scrapy/http/request/form.py [hunk 1]
```diff
@@ -170,10 +170,12 @@ def _get_clickable(clickdata, form):
     """
     clickables = [
         el for el in form.xpath(
-            'descendant::*[(self::input or self::button)'
-            ' and re:test(@type, "^submit$", "i")]'
-            '|descendant::button[not(@type)]',
-            namespaces={"re": "http://exslt.org/regular-expressions"})
```
Comment (dangra) [inline scrapy/http/request/form.py]: is it me or these lines are changed from 4 spaces indentation to 5 spaces?

Comment (dangra) [reply scrapy/http/request/form.py]: TL;DR it would be much easier to review if you keep the indentation at 4 spaces.

Comment (virmht) [reply scrapy/http/request/form.py]: Sure, I will make the necessary changes.

Comment (virmht) [reply scrapy/http/request/form.py]: @dangra @cathalgarvey Resolved this review. Please see the new changes.

File: scrapy/http/request/form.py [hunk 2]
```diff
@@ -172,6 +172,8 @@ def _get_clickable(clickdata, form):
         el for el in form.xpath(
             'descendant::*[(self::input or self::button)'
             ' and re:test(@type, "^submit$", "i")]'
+            '|descendant::*[(self::input or self::button)'
+            ' and re:test(@type, "^image$", "i")]'
```
Comment (dangra) [inline scrapy/http/request/form.py]: Considering `button[type=image]` isn't valid, I would go for:

```
descendant::input[re.test(@type, "^(submit|image)$", "i")]
|descendant::button[not(@type) or re.test(@type, "^submit$", "i")]
```

Comment (virmht) [reply scrapy/http/request/form.py]: Done! @dangra, please see the new changes.

**Comment by virmht** [comment]
@dangra, @cathalgarvey, - suggestions are welcome :)

**Comment by dangra** [comment]
thanks @virmht !

**Pull request merged**